ikarus/io/vtkwriter.hh

```diff
@@ -0,0 +1,182 @@
+// SPDX-FileCopyrightText: 2021-2024 The Ikarus Developers mueller@ibb.uni-stuttgart.de
+// SPDX-License-Identifier: LGPL-3.0-or-later
+
+/**
+ * \file vtkwriter.hh
+ * \brief Ikarus VTK Writer for finite element results
+ * \ingroup io
+ *
+ */
+
+#pragma once
+
+#include <dune/functions/gridfunctions/discreteglobalbasisfunction.hh>
+#include <dune/vtk/vtkwriter.hh>
+#include <dune/vtk/writers/unstructuredgridwriter.hh>
+
+#include <ikarus/io/resultfunction.hh>
+#include <ikarus/utils/concepts.hh>
+
+namespace Ikarus::Concepts {
+
+// adapted from /dune/dune-vtk/dune/vtk/utility/concepts.hh
+template <class DC>
+concept IsDataCollector = requires(DC dc) {
+  typename DC::GridView;
+  dc.update();
+  dc.numPoints();
+  dc.numCells();
+};
+} // namespace Ikarus::Concepts
+
+namespace Ikarus::Vtk {
+
+namespace Impl {
+  struct AsCellData
+  {
+  };
+  struct AsPointData
+  {
+  };
+  namespace Concepts {
+    template <typename DT>
+    concept DataType = std::is_same_v<DT, AsCellData> or std::is_same_v<DT, AsPointData>;
+  }
+} // namespace Impl
+
+template <typename AS, bool structured = false, typename DC = void>
+requires(Concepts::FlatAssembler<AS> and (std::is_same_v<DC, void> or Concepts::IsDataCollector<DC>))
+class Writer
+{
+public:
+  using Assembler     = AS;
+  using GridView      = typename Assembler::GridView;
+  using FERequirement = typename Assembler::FERequirement;
+  using FEContainer   = typename Assembler::FEContainerType;
+  using FEType        = typename std::remove_cvref_t<FEContainer>::value_type;
+
+  // We are using the provided DataCollector, but if none was provided we are using the default ones from dune-vtk
+  using DataCollector = std::conditional_t<std::is_same_v<DC, void>,
+                                           std::conditional_t<structured, Dune::Vtk::YaspDataCollector<GridView>,
+                                                              Dune::Vtk::ContinuousDataCollector<GridView>>,
+                                           DC>;
+
+  // We are using a RectilinearGridWriter if structured is true
+  using UnderlyingVTKWriter = std::conditional_t<structured, Dune::Vtk::RectilinearGridWriter<GridView, DataCollector>,
```

Comment:
Same for `DefaultVTKWriter` as above.

Reply:
additionall this class implementation will never have to be touched even if somebody wants to do something fancy